test/benchee/benchmark/runner_test.exs

```diff
@@ -611,97 +734,124 @@ defmodule Benchee.Benchmark.RunnerTest do
-      ]
+      assert_received_exactly([
+        {:global_scenario, 0},
+        {:local_scenario, 1},
+        {:global_each, 2},
+        {:local_each, 3},
+        {:runner, 4},
+        {:global_each, 2},
+        {:local_each, 3},
+        {:runner, 4},
+        {:local_after_scenario, 2},
+        {:global_after_scenario, 2}
+      ])
     end
 
     test "hooks dealing with inputs still work when there is no input given" do
       me = self()
+
       %Suite{
         configuration: %{
           warmup: 1,
           time: 1,
-          before_scenario: fn(input) ->
+          before_scenario: fn input ->
             send(me, {:global_scenario, input})
             input
           end,
-          before_each: fn(input) ->
+          before_each: fn input ->
             send(me, {:global_each, input})
             input
           end,
-          after_scenario: fn(input) ->
+          after_scenario: fn input ->
             send(me, {:global_after_scenario, input})
           end
         }
       }
       |> test_suite
       |> Benchmark.benchmark("job", {fn ->
-           # still keep to make sure we only get one iteration and not too fast
-           :timer.sleep 1
-         end,
-         before_scenario: fn(input) ->
-           send(me, {:local_scenario, input})
-           input
-         end,
-         before_each: fn(input) ->
-           send(me, {:local_each, input})
-           input
-         end,
-         after_scenario: fn(input) ->
-           send(me, {:local_after_scenario, input})
-         end})
+         # still keep to make sure we only get one iteration and not too fast
+         :timer.sleep(1)
+       end,
+       before_scenario: fn input ->
+         send(me, {:local_scenario, input})
+         input
+       end,
+       before_each: fn input ->
+         send(me, {:local_each, input})
+         input
+       end,
+       after_scenario: fn input ->
+         send(me, {:local_after_scenario, input})
+       end})
       |> Benchmark.measure(TestPrinter)
 
       no_input = Benchmark.no_input()
 
-      assert_received_exactly [
-        {:global_scenario, no_input}, {:local_scenario, no_input},
-        {:global_each, no_input}, {:local_each, no_input},
-        {:global_each, no_input}, {:local_each, no_input},
-        {:local_after_scenario, no_input}, {:global_after_scenario, no_input}
-      ]
+      assert_received_exactly([
+        {:global_scenario, no_input},
+        {:local_scenario, no_input},
+        {:global_each, no_input},
+        {:local_each, no_input},
+        {:global_each, no_input},
+        {:local_each, no_input},
+        {:local_after_scenario, no_input},
+        {:global_after_scenario, no_input}
+      ])
     end
 
+    test "runs all benchmarks with all inputs exactly once as a dry run" do
+      ref = self()
+
+      inputs = %{"small" => 1, "big" => 100}
+
+      config = %{time: 0, warmup: 0, inputs: inputs, dry_run: true}
```

Comment:
if we wanna catch what I think I identified as a bug above (dry run should run completely before anything "real" is run) then we could give it warmup and time and make the last function crash (or add a function that crashes). Then we can assert that exactly these ones were received.

Reply:
You know what - I'll put an error in the `after_scenario` and make sure we only get the `before_each` and `after_each` hooks once. That should do it!